Document asn.1 encoding functions (#210)
smallstep/docs: step-ca/templates.mdx

```
"extensions": [
  {"id": "1.2.3.4", "critical": false, "value": {{ asn1Enc "int:3848281" | toJson }}}
]
```

When applied to template variables, these functions enable dynamic OID extensions:

```
{
	"subject": {{ toJson .Subject }},
	"sans": {{ toJson .SANs }},
{{- if typeIs "*rsa.PublicKey" .Insecure.CR.PublicKey }}
	"keyUsage": ["keyEncipherment", "digitalSignature"],
{{- else }}
	"keyUsage": ["digitalSignature"],
{{- end }}
	"extKeyUsage": ["serverAuth", "clientAuth"],
	"extensions": [
		{"id": "1.2.3.4", "value": {{ asn1Marshal .AuthorizationCrt.NotAfter | toJson }}},
		{"id": "1.2.3.5", "value": {{ asn1Set (asn1Marshal (first .Insecure.CR.DNSNames) "utf8") (asn1Enc "int:123456") | toJson }}},
		{"id": "1.2.3.6", "value": {{ asn1Seq (asn1Enc "YubiKey") (asn1Enc "int:123456") | toJson }}}
	]
}
```

Let's walk through one line of this template:

```
{"id": "1.2.3.6", "value": {{ asn1Seq (asn1Enc "YubiKey") (asn1Enc "int:123456") | toJson }}}
```

To render this line of the template:

1. `asn1Enc "YubiKey"` encodes "YubiKey" as an ASN.1 Printable string, and returns the Base64
2. `asn1Enc "int:123456"` encodes 123 as an ASN.1 Integer, and returns the Base64

> 3. `asn1Seq` gets the resulting Base64 strings, decodes them, creates an ASN.1 SEQENCE, and returns the Base64 of the SEQUENCE.

Reviewer: ```suggestion
3. `asn1Seq` gets the resulting Base64 strings, decodes them, creates an ASN.1 SEQUENCE, and returns the Base64 of the SEQUENCE.
```

4. Finally, `toJson` returns a JSON string representation of the Base64

#### asn1Enc

Encodes a string into an ASN.1 value.

A data type can be supplied as a prefix,
and ASN.1 Printable is the default data type:

```
asn1Enc "int:123"
asn1Enc "oid:1.2.3.4"
asn1Enc "foo"
asn1Enc .Subject.CommonName
```

This function supports the data types described [here](https://github.com/smallstep/crypto/blob/878092791d7e5ef8530dfa949cfe3cde61373d77/x509util/extensions.go#L486):

- **printable**: "printable:abcfoo" (or just "abcfoo")
- **int**: "int:12324"
- **oid**: "oid:1.2.3.4"
- **utf8**: "utf8:foobar"
- **ia5**: "ia5:abcfoo"

> - **numeric**: "numeric:01 02 03345" (multiple space-separated ints)

Reviewer: It allows numbers and spaces, you can also use numeric:1234, you don't have to use spaces.

from the `"templateData"` object
in your provisioner's [configuration block](#configuring-the-ca-to-use-templates).

### ASN.1 Values

The X.509 templating system includes functions for ASN.1 encoding in a template.
This can be useful for populating custom OID `extensions`:

```
"extensions": [

>   {"id": "1.2.3.4", "critical": true, "value": {{ asn1Enc "int:3848281" | toJson }}}

Reviewer: Usually critical will be `false` on custom extensions. If the client/server does not understand that extension it should fail.
```suggestion
  {"id": "1.2.3.4", "critical": false, "value": {{ asn1Enc "int:3848281" | toJson }}}
```

```
{
	"subject": {{ toJson .Subject }},
	"sans": {{ toJson .SANs }},
{{- if typeIs "*rsa.PublicKey" .Insecure.CR.PublicKey }}
	"keyUsage": ["keyEncipherment", "digitalSignature"],
{{- else }}
	"keyUsage": ["digitalSignature"],
{{- end }}
	"extKeyUsage": ["serverAuth", "clientAuth"],
	"extensions": [
		{"id": "1.2.3.5", "value": {{ asn1Marshal .AuthorizationCrt.NotAfter | toJson }}},
		{"id": "1.2.3.6", "value": {{ asn1Seq (asn1Enc (first .Insecure.CR.DNSNames)) (asn1Enc "int:123456") | toJson }}},
		{"id": "1.2.3.7", "value": {{ asn1Set (asn1Marshal (first .Insecure.CR.DNSNames) "utf8") (asn1Enc "int:123456") | toJson }}}
	]
}
```

All of the functions here return base64.

To make that possible, `asn1Seq` and `asn1Set` will decode a list of base64 values before putting them into an ASN.1 set or sequence, which is then re-encoded to base64.

For example:

```
{"id": "1.2.3.4", "value": {{ asn1Seq (asn1Enc "YubiKey") (asn1Enc "int:123456") | toJson }}}
```

1. `asn1Enc "YubiKey"` will encode "YubiKey" as an ASN.1 Printable string, and return the base64
2. `asn1Enc "int:123456"` will encode 123 as an ASN.1 Integer, and return the base64

> 3. `ans1Seq` wil get all the base64 strings, decode them, create an ASN.1 SEQENCE and return the base64 of the SEQUENCE.

Reviewer: ```suggestion
3. `asn1Seq` will get all the base64 strings, decode them, create an ASN.1 SEQENCE and return the base64 of the SEQUENCE.
```

> 4. Finally, `toJson` returns a JSON string representation of the base64

Reviewer: `asn1Set` is not described.

#### asn1Enc

Encodes a string into an ASN.1 value.

> A data type can be supplied inline.

Reviewer: It supports the types described in [this function](https://github.com/smallstep/crypto/blob/878092791d7e5ef8530dfa949cfe3cde61373d77/x509util/extensions.go#L486):
* int: "int:1234"
* oid: "oid:1.2.3.4"
* utf8: "utf8:string"
* ia5: "ia5:srtring"
* numeric: "numeric:1234"
* printable: "printable:string" or just "string"
* utc: "utc:2023-03-29T02:03:57Z" or "utc:2023-04-20 15:28:27.909088 -0700 PDT"
* generalized: "generalized:2023-03-29T02:03:57Z" or "generalized:2023-04-20 15:28:27.909088 -0700 PDT"
* raw: a special one that allows any base64 string

```
asn1Enc "int:123"
asn1Enc "foo"
asn1Enc .Subject.CommonName
```

#### asn1Seq

Creates an ASN.1 `SEQUENCE` of values.

```
asn1Seq (asn1Enc "foo") (asn1Enc "int:123") ...
```

#### asn1Set

Creates an ASN.1 `SET` of values.

```
asn1Set (asn1Enc "foo") (asn1Enc "int:123") ...
```

#### asn1Marshal

The intention of `asn1Marshal` is to be able to marshal variable data.
For example, if you want to encode the Not After value in an X5C authorization certificate,
without converting it to a string first, `asn1Marshal .AuthorizationCrt.NotAfter` will do the trick.

`asn1Marshal` imitates Go's [`asn1.MarshalWithParams`](https://pkg.go.dev/encoding/asn1#MarshalWithParams).

```
asn1Marshal .Token.iss "printable"
asn1Marshal .AuthorizationCrt.NotAfter "utc"
```

The second parameter ("printable", "utc") is the ASN.1 data type.

> The data types are listed in the [`encoding/asn1` package documentation](https://pkg.go.dev/encoding/asn1)

Reviewer: The second parameter is not required and it will use Go's default transformation.
